### PR TITLE
[GEOT-5602] Better concurrent SoftValueHashMap

### PR DESCRIPTION
**Pull request by ugomos**
SoftValueHashMap now supports concurrency better, using ConcurrentHashMap and ConcurrentLinkedQueue.

### CODE REVIEW

File: modules/library/metadata/src/main/java/org/geotools/util/SoftValueHashMap.java [hunk 1]
```diff
@@ -44,11 +45,11 @@
  * collection. The amount of entries to retain by hard reference is specified at {@linkplain
  * #SoftValueHashMap(int) construction time}.
  * <p>
- * This map is thread-safe. It accepts the null key, but doesn't accepts null values. Usage
- * of {@linkplain #values value}, {@linkplain #keySet key} or {@linkplain #entrySet entry}
- * collections are supported except for direct usage of their iterators, which may throw
- * {@link java.util.ConcurrentModificationException} randomly depending on the garbage collector
- * activity.
+ * This map is thread-safe. Since it uses the concurrent structures {@linkplain ConcurrentHashMap}
+ * and {@linkplain ConcurrentLinkedQueue}, such structures do not accept either the null key and null values, and
```
Comment (aaime) [inline modules/library/metadata/src/main/java/org/geotools/util/SoftValueHashMap.java]: The usage of the placeholder should be a internal detail, no need to tell clients about it. Or do they need to know for any reason?

File: modules/library/metadata/src/main/java/org/geotools/util/SoftValueHashMap.java [hunk 2]
```diff
@@ -204,67 +200,64 @@ public boolean containsValue(final Object value) {
...
+             * If the value is a soft reference, get the referent and clear it immediately
+             * for avoiding the reference to be enqueued. We abandon the soft reference and
+             * reinject the referent as a strong reference in the hash map, since we try to
+             * keep the last entries by strong references.
              */
-            @SuppressWarnings("unchecked")
-            final V v = (V) value;
-            return v;
-        }
+             value = ((Reference<K, V>) value).getAndClear();
```
Comment (aaime) [inline modules/library/metadata/src/main/java/org/geotools/util/SoftValueHashMap.java]: Oh, interesting... but this means we need to only clear the soft reference, but not actually call the cleaner as the object is not being destroyed

File: modules/library/metadata/src/main/java/org/geotools/util/SoftValueHashMap.java [hunk 3]
```diff
@@ -562,13 +511,12 @@ public String toString() {
...
         private boolean findNext() {
-            assert Thread.holdsLock(hash);
             while (iterator.hasNext()) {
                 final Map.Entry<K,Object> candidate = iterator.next();
                 Object value = candidate.getValue();
                 if (value instanceof Reference) {
-                    value = ((Reference) value).get();
-                    entry = new MapEntry<K,V>(candidate.getKey(), (V) value);
+                    value = ((Reference<K, V>) value).get();
+                    entry = new MapEntry<K,V>((K) SoftValueHashMap.replaceNull(candidate.getKey()), (V) value);
```
Comment (aaime) [inline modules/library/metadata/src/main/java/org/geotools/util/SoftValueHashMap.java]: Shouldn't this be the other way around? if the placeholder is used, replace it with null instead? Is there a test to cover this, if not, please add one.

File: modules/library/metadata/src/main/java/org/geotools/util/SoftValueHashMap.java [hunk 4]
```diff
@@ -660,9 +602,8 @@ public void remove() {
          * or is about to be removed.</li>
          */
         final Object getAndClear() {
-            assert Thread.holdsLock(hash);
             final Object value = get();
-            super.clear();
+            clear();
```
Comment (aaime) [inline modules/library/metadata/src/main/java/org/geotools/util/SoftValueHashMap.java]: See my comment above, it's likely that the original code was correct, cleaning the reference without running the cleaner.

**Comment by aaime** [comment]
Tested in a benchmark setting, it indeed improves scalability

**Comment by bencaradocdavies** [comment]
Looks like this PR causes intermittent build failures in GeoServer gs-wms:
http://osgeo-org.1560.x6.nabble.com/Random-WMS-failure-on-build-box-td5303350.html

**Comment by aaime** [comment]
Reverting as it's causing build failures in GeoServer. The full stack trace is:

````
14 Jan 11:17:43 ERROR [geotools.rendering] - org.geotools.util.SoftValueHashMap$Reference cannot be cast to org.opengis.referencing.operation.CoordinateOperation
java.lang.ClassCastException: org.geotools.util.SoftValueHashMap$Reference cannot be cast to org.opengis.referencing.operation.CoordinateOperation
	at org.geotools.referencing.operation.BufferedCoordinateOperationFactory.createOperation(BufferedCoordinateOperationFactory.java:256)
	at org.geotools.geometry.jts.ReferencedEnvelope.transform(ReferencedEnvelope.java:688)
	at org.geotools.geometry.jts.ReferencedEnvelope.transform(ReferencedEnvelope.java:635)
	at org.geotools.renderer.crs.ProjectionHandler.preProcess(ProjectionHandler.java:417)
	at org.geotools.renderer.lite.StreamingRenderer$RenderableFeature.getTransformedShape(StreamingRenderer.java:3117)
	at org.geotools.renderer.lite.StreamingRenderer$RenderableFeature.getShape(StreamingRenderer.java:3077)
	at org.geotools.renderer.lite.StreamingRenderer.processSymbolizers(StreamingRenderer.java:2549)
	at org.geotools.renderer.lite.StreamingRenderer.processFeature(StreamingRenderer.java:2449)
	at org.geotools.renderer.lite.StreamingRenderer.drawPlain(StreamingRenderer.java:2306)
	at org.geotools.renderer.lite.StreamingRenderer.processStylers(StreamingRenderer.java:1927)
	at org.geotools.renderer.lite.StreamingRenderer.paint(StreamingRenderer.java:831)
	at org.geoserver.wms.map.RenderedImageMapOutputFormat.produceMap(RenderedImageMapOutputFormat.java:545)
	at org.geoserver.wms.map.RenderedImageMapOutputFormat.produceMap(RenderedImageMapOutputFormat.java:268)
	at org.geoserver.wms.map.RenderedImageMapOutputFormat.produceMap(RenderedImageMapOutputFormat.java:138)
	at org.geoserver.wms.GetMap.executeInternal(GetMap.java:623)
	at org.geoserver.wms.GetMap.run(GetMap.java:279)
	at org.geoserver.wms.GetMap.run(GetMap.java:125)
	at org.geoserver.wms.DefaultWebMapService.getMap(DefaultWebMapService.java:320)
	at sun.reflect.GeneratedMethodAccessor682.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:302)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.geoserver.ows.util.RequestObjectLogger.invoke(RequestObjectLogger.java:33)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
	at com.sun.proxy.$Proxy44.getMap(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor683.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.geoserver.ows.Dispatcher.execute(Dispatcher.java:850)
	at org.geoserver.ows.Dispatcher.handleRequestInternal(Dispatcher.java:268)
	at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:147)
	at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:50)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:959)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:893)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:968)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:859)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:668)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:844)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:770)
	at org.geoserver.test.GeoServerSystemTestSupport$1.service(GeoServerSystemTestSupport.java:1478)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:770)
	at org.springframework.mock.web.MockFilterChain$ServletFilterProxy.doFilter(MockFilterChain.java:167)
	at org.springframework.mock.web.MockFilterChain.doFilter(MockFilterChain.java:134)
	at org.geoserver.test.GeoServerSystemTestSupport.dispatch(GeoServerSystemTestSupport.java:1505)
	at org.geoserver.test.GeoServerSystemTestSupport.dispatch(GeoServerSystemTestSupport.java:1436)
	at org.geoserver.test.GeoServerSystemTestSupport.getAsServletResponse(GeoServerSystemTestSupport.java:977)
	at org.geoserver.test.GeoServerSystemTestSupport.getAsServletResponse(GeoServerSystemTestSupport.java:960)
	at org.geoserver.test.GeoServerSystemTestSupport.getAsImage(GeoServerSystemTestSupport.java:1248)
	at org.geoserver.wms.wms_1_1_1.GetMapIntegrationTest.testAdvancedProjectionHandling(GetMapIntegrationTest.java:1094)
````

**Pull request merged**